Support work.result() to get result tensors for allreduce for Gloo, NCCL backends

mrzzd opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43386 Support work.result() to get result tensors for allreduce for Gloo, NCCL backends**

Differential Revision: [D23216393](https://our.internmc.facebook.com/intern/diff/D23216393/)

Commit:
Support work.result() to get result tensors for allreduce for Gloo, NCCL backends



Differential Revision: [D23216393](https://our.internmc.facebook.com/intern/diff/D23216393/)

[ghstack-poisoned]

mrshenli reviewed (commented):
Hey @mrzzd, could you please explain the motivation of adding this API in the PR and diff summary? thanks!

mrzzd commented:
> Hey @mrzzd, could you please explain the motivation of adding this API in the PR and diff summary? thanks!

Please see #43178 for more context.

mrshenli commented:
> Please see #43178 for more context.

Thanks. Shall we add that to the diff/PR description, so that it is included in the git commit message?

mrzzd commented:
> > Please see #43178 for more context.
> 
> Thanks. Shall we add that to the diff/PR description, so that it is included in the git commit message?

Added it to diff description, but does not show up here. Anyhow, the issue is linked to the PR. So it should show in merge commit description, does not  it?

mrshenli reviewed (commented):
Since the `result` API is also exposed to Python, let's also add some Python tests. 

also adding @rohan-varma for review, as this feature might be used by #42577

rohan-varma commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -818,6 +820,14 @@ class AsyncAllreduceWork : public ProcessGroupGloo::AsyncWork {
   }
+
+
+  std::vector<at::Tensor> result() const override {
```
What happens if the result() is not ready (such as operation was ran asynchronously?) We should probably have a check for this, otherwise there may be a risk of getting back incorrect results. 

We can probably do something like `TORCH_CHECK(isCompleted())` and throw an appropriate error message if it is not.

mrzzd replied:
That's a good point, done. PTAL.

rohan-varma reviewed (commented):
Thanks for working on this! As @mrshenli mentioned, we would need Python tests since `result()` will probably be mostly accessed by users in Python (I should've clarified this in the issue description). In particular, we should probably test the following for both NCCL and Gloo (probably with CPU and GPU tensors for Gloo as well):
1) Call dist.all_reduce() API with `async_op = True` 
2) Check that `result()` throws since we have not waited
3) Now call `wait()` and ensure `result()` gives back the expected tensors. 
4) Call `dist.all_reduce()` synchronously and ensure `result()` works. 

Also, since this is an OSS PR it would be great to have a description of the change that lives in the GitHub PR body so that future developers can have context.

Commit:
Update on "Support work.result() to get result tensors for allreduce for Gloo, NCCL backends "


Differential Revision: [D23216393](https://our.internmc.facebook.com/intern/diff/D23216393/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/test_distributed.py:
```diff
@@ -847,6 +847,43 @@ def test_reduce_full_group_max(self):
+            else:
+                work = group_id.allreduce([tensor], opts)
+
+            # Calling result right the work is finished should throw exception.
```
Is there a better way to guarantee under some scenario that this async work is not completed after calling into the API? (maybe @mrshenli may have some suggestions). If not, then this try-except approach looks like the best we can do.

mrzzd replied:
Not sure. In the case of Gloo, almost always we run into exception, but for NCCL, it is too fast, and very likely that the exception does not happend.

rohan-varma commented on test/distributed/test_distributed.py:
```diff
@@ -847,6 +847,43 @@ def test_reduce_full_group_max(self):
+                work.result()
+            except RuntimeError:
+                throws_exception = True
+            EXPECT_TRUE(throws_exception or work.is_completed())
```
Probably a more canonical way to do this is:

```
try:
    with self.assertRaisesRegex(RuntimeErorr, expected_message_str(:
        work.result()
except:
    # Exception was not raised, ensure is_completed()
    self.assertTrue(work.is_completed())
```

mrzzd replied:
Actually we want to test "work.is_completed()" when exception is NOT raised

rohan-varma replied:
Yes, I believe that is what my code block does unless I'm missing something?
In the code block, 
1) If work.result() raises exception, it is caught by `assertRaisesRegex` and we validate the exception appropriately. There is no further checking after that.
2) if work.result() does NOT raise an exception, then `assertRaisesRegex` should itself raise an exception indicating no exception was raised. This exception is caught in the `except` block, wherein we validate `work.is_completed()` as we expect (since in this branch work.result() should have returned without errors).

mrzzd replied:
No you did not miss something, I read your code snippet too fast. Fixed, thanks!

rohan-varma commented on test/distributed/test_distributed.py:
```diff
@@ -847,6 +847,43 @@ def test_reduce_full_group_max(self):
+                work.result()
+            except RuntimeError:
+                throws_exception = True
+            EXPECT_TRUE(throws_exception or work.is_completed())
```
In python, this should be `self.assertTrue` not `EXPECT_TRUE` right?

mrzzd replied:
my bad, last minute change :) fixed.

Commit:
Update on "Support work.result() to get result tensors for allreduce for Gloo, NCCL backends "


Differential Revision: [D23216393](https://our.internmc.facebook.com/intern/diff/D23216393/)

[ghstack-poisoned]

Commit:
Update on "Support work.result() to get result tensors for allreduce for Gloo, NCCL backends "


Differential Revision: [D23216393](https://our.internmc.facebook.com/intern/diff/D23216393/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM, thanks! Please ensure all CI tests are green before landing

rohan-varma commented on test/distributed/test_distributed.py:
```diff
@@ -847,6 +847,45 @@ def test_reduce_full_group_max(self):
+
+            work.wait()
+
+            result = work.result()
```
My bad, should have noticed this earlier. I think this can fail because we check `isCompleted()`, but for ProcessGroupNCCL `isCompleted()` can still be false after a call to `wait()` due to CUDA async model.
